llvm/lib/Target/AMDGPU/AMDGPUNextUseAnalysis.cpp

```diff
@@ -0,0 +1,573 @@
+//===---------------------- AMDGPUNextUseAnalysis.cpp  --------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "AMDGPUNextUseAnalysis.h"
+#include "AMDGPU.h"
+#include "GCNSubtarget.h"
+#include "llvm/InitializePasses.h"
+#include <limits>
+#include <queue>
+
+using namespace llvm;
+
+#define DEBUG_TYPE "amdgpu-next-use-analysis"
+
+static cl::opt<bool> DumpNextUseDistance("dump-next-use-distance",
+                                         cl::init(false), cl::Hidden);
+
+bool AMDGPUNextUseAnalysis::isBackedge(MachineBasicBlock *From,
+                                       MachineBasicBlock *To) const {
+  MachineLoop *Loop1 = MLI->getLoopFor(From);
+  MachineLoop *Loop2 = MLI->getLoopFor(To);
+  if (!Loop1 || !Loop2 || Loop1 != Loop2)
+    return false;
+  SmallVector<MachineBasicBlock *, 2> Latches;
+  Loop1->getLoopLatches(Latches);
+  auto It = llvm::find(Latches, From);
+  MachineBasicBlock *LoopHeader = Loop1->getHeader();
+  return It != Latches.end() && From->isSuccessor(To) && To == LoopHeader;
```

Comment:
You can check `isSuccessor` right at the top of the function and exit early if it isn't. Likewise, you can check if To is Loop1's header before getting the loop latches and calling find on them.